posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,218 @@
+
+```lean
+example : (if 37 * 0 then 1 else 2) = 1 := by
+  -- Works since `simp` can simplify `37 * 0` to `True`
+  -- because it knows the lemma `mul_zero a : a * 0 = 0`.
+  simp only [↓reduceIte]
+
+example (X : Type) (P : Prop) (a b : X) : (if P ∨ ¬ P then a else b) = a := by
+  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
+  simp only [↓reduceIte]
+
+open Classical
+
+example : (if FermatLastTheorem then 1 else 2) = 1 := by
+  --This fails because `simp` can't simplify `FermatLastTheorem` to `True` or `False`
+  simp only [↓reduceIte]
+  -- Remaining goal: `⊢ (if FermatLastTheorem then 1 else 2) = 1`
+  sorry -- See https://imperialcollegelondon.github.io/FLT for how to solve this `sorry` ;)
+```
+
+The down arrow `↓` in `simp only [↓reduceIte]` indicates that `reduceIte` is being called as a *preprocedure*.
+This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.
+
+> Recall that we said that, as a simple approximation, simplification is performed from the inside-out.
+> This is a concrete example where simplification happens outside-in.
+
+Concretely, `simp` needs only simplify one of the expressions `a` and `b`, as the other one is discarded upfront.
+For example, `↓reduceIte` allows `simp` to avoid touching `bigComplicatedExpression` at all in the following:
+```
+example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
+  simp only [↓reduceIte]
+```
+
+Notice that `simp [↓reduceIte]` is equivalent to `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter simplifies inside both branches of the `ite`, instead of simplifying away the irrelevant one before passing to subexpressions.
+
+When presented with a left hand side of the form `ite P a b`, `reduceIte` does the following:
+- Call `simp` on `P` to get a simplified expression `P'` and a proof `h` that `P = P'`.
+- If `P'` is `True` then return the right hand side `a` together with the proof `ite_cond_eq_true r` that `ite P a b = a`.
+- If `P'` is `False` then return the the right hand side `b` togethr with the proof `ite_cond_eq_false r` that `ite P a b = b`.
+
+## Many more applications!
+
+In the second blog post, we will see how to build step by step a simproc for computing a variant of `List.range` when the parameter is a numeral.
+
+# A few caveats
+
+The current design of simprocs comes with a few restrictions that are worth keeping in mind:
+* By definition, **a simproc can only be used in `simp`** (and `simp`-like tactics like `simp_rw`, `simpa`, `aesop`), even though the notion of a "modular lemma" could be useful in other rewriting tactics like `rw`.
+* **One cannot provide arguments to a simproc to restrict the occurrences it rewrites**.
+  In contrast, this is possible for lemmas in all rewriting tactics: eg `rw [add_comm c]` turns `⊢ a + b = c + d` into `⊢ a + b = d + c` where `rw [add_comm]` would instead have turned it into `⊢ b + a = c + d`.
```

Comment:
I was hoping this would work, but alas:
```lean
def myProc (a : Nat) : Lean.Meta.Simp.Simproc := fun e => do
  return .continue

-- illegal
simproc_pattern% _ => myProc 

-- ok, but arguments were consumed already
def wat := myProc 1
simproc_pattern% _ => wat
```

Reply:
Yes sadly I think a few things to change for Lean to acquire `rwproc`